[Serverless][Security Solution][Endpoint] Gate endpoint exceptions on rule details and API changes

ashokaditya opened the pull request:
## What this PR changes

Follow up of elastic/kibana/pull/164107/

For serverless ES/Kibana, it gates exception list API for endpoint exceptions and restricts endpoint exceptions tab on Endpoint Security rule details based on project PLIs. If no endpoint PLIs, endpoint exceptions should not be accessible.

- [x] Add upselling to `app/security/exceptions/details/endpoint_list` page
- [ ] Tests (WIP) - in a follow up PR

### How to review

Best to follow along commits for a code review. Below are details to manually test the changes.

- Setup for _Servlerless_
  - Run `yarn es serverless --kill --clean --license trial -E xpack.security.authc.api_key.enabled=true` on a terminal window to start ES.
  - Copy `config/serverless.security.yml` to `config/serverless.security.dev.yml` 
  - Run `yarn serverless-security --no-base-path` on another terminal window to start kibana in serverless mode
  - Log in using `serverless_security` user.

### Tests (Serverless)
This needs to be tested with a custom user/role and not `elastic_serverless` which has `superuser` role.

1. ### PLI configs
`{ product_line: 'security', product_tier: 'essentials' }` or `{ product_line: 'security', product_tier: 'complete' }` 
and
`{ product_line: 'endpoint', product_tier: 'essentials' }` or `{ product_line: 'endpoint', product_tier: 'complete' }`

- #### UX
  1.  Navigate to Rules via `http://localhost:5601/app/security/rules/`. Click on `Add Elastic rules`.
  2. Select and add `Endpoint Security` rule.
  3. Click `Endpoint Security` and navigate to the rules details page, and you should see `Endpoint exceptions` tab. The tabs visible are `Alerts`, `Endpoint exceptions`, `Rule exceptions`, `Execution results`.
  4. Navigate to Rules>Shared Exception Lists > Endpoint Security Exception List via `app/security/exceptions/details/endpoint_list` and you should be able to see the page with any added endpoint exceptions.

- #### API requests (with user `serverless_security`)
  1. should get a status `200` on`POST api/exception_lists/items`
  2. should get a status `200` on `POST api/exception_lists/_export?id=endpoint_list&list_id=endpoint_list&namespace_type=agnostic&include_expired_exceptions=true`
  3. should get a status `200` on `PUT api/exception_lists/items`
  5. should get a status `200` on `DELETE api/exception_lists/items`
  6. should get a status `200` on `GET api/exception_lists/items/_find?list_id=endpoint_list&namespace_type=agnostic`

2. ### PLI configs
`{ product_line: 'security', product_tier: 'essentials' }` or `{ product_line: 'security', product_tier: 'complete' }`

- #### UX
  1.  Navigate to Rules via `http://localhost:5601/app/security/rules/`. Click on `Add Elastic rules`.
  2. Select and add `Endpoint Security` rule. 
  3. Click `Endpoint Security` and navigate to the rules details page, and you should not see `Endpoint exceptions` tab. The only tabs visible are `Alerts`, `Rule exceptions`, `Execution results`.   
   ![Screenshot 2023-09-14 at 3 33 24 PM](https://github.com/elastic/kibana/assets/1849116/185ea210-c457-4469-a824-cdcaa2893cb6)
  4. Navigate to Rules>Shared Exception Lists > Endpoint Security Exception List via `app/security/exceptions/details/endpoint_list` and you should see an upsell message.  
![Screenshot 2023-09-14 at 3 29 14 PM](https://github.com/elastic/kibana/assets/1849116/6700fc2d-9a9d-4a57-ad7f-5505e02cec71)


- #### API requests
  1. should get a status `403` on`POST api/exception_lists/items`
  2. should get a status `403` on `POST api/exception_lists/_export?id=endpoint_list&list_id=endpoint_list&namespace_type=agnostic&include_expired_exceptions=true`
  3. should get a status `403` on `PUT api/exception_lists/items`
  6. should get a status `403` on `DELETE api/exception_lists/items`
  7. should get a status `403` on `GET api/exception_lists/items/_find?list_id=endpoint_list&namespace_type=agnostic`
---


**Flaky FTRs**
https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3248
https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/3255


### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

ashokaditya committed:
gate endpoint exceptions on rule details

refs elastic/security-team/issues/7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya commented:
@elasticmachin a merge upstream

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
add endpoint exceptions privileges to authz

refs elastic/security-team/issues/7223

ashokaditya committed:
add endpoint exceptions validator

refs elastic/security-team/issues/7223

ashokaditya committed:
use endpoint exceptions validator in extensions

refs elastic/security-team/issues/7223

ashokaditya committed:
update initial authz state in test

ashokaditya committed:
Update authz.test.ts

ashokaditya committed:
update error class for endpoint exceptions

refs elastic/security-team/issues/7223

ashokaditya committed:
use list id const

refs elastic/security-team/issues/7223

ashokaditya committed:
fix type imports

refs elastic/kibana/pull/164107

ashokaditya committed:
fix type import

related to elastic/kibana/pull/163406

ashokaditya committed:
show upsell page for exception details

refs elastic/security-team/issues/7223

elasticmachine commented:
Pinging @elastic/security-defend-workflows (Team:Defend Workflows)

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

ashokaditya committed:
upselling page related capability

refs elastic/security-team/issues/7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya commented on x-pack/plugins/security_solution_serverless/public/upselling/pages/endpoint_exceptions_details.tsx:
```diff
@@ -0,0 +1,47 @@
+          <h2>
+            <FormattedMessage
+              id="xpack.securitySolutionServerless.endpoint.exceptions.details.paywall.title"
+              defaultMessage="Do more with Security!"
+            />
+          </h2>
+        }
+        body={
+          <p>
+            <FormattedMessage
+              id="xpack.securitySolutionServerless.endpoint.exceptions.details.paywall.body"
+              defaultMessage="Upgrade your license to {productTypeRequired} to use Endpoint Security Exception List."
```
@caitlinbetz @kevinlog The info message needs an update here.

paul-tavares replied:
Will we (our team) continue to be the ones responsible for these files? if so, can I ask that you create a folder (maybe name it the same as the one I created under `sections/` - `endpoint_management`) and update the `CODEOWNERS` file to show our team as owners of it?

paul-tavares commented on x-pack/plugins/security_solution/common/endpoint/service/authz/authz.ts:
```diff
@@ -30,6 +30,13 @@ export function hasKibanaPrivilege(
   return fleetAuthz.packagePrivileges?.endpoint?.actions[privilege].executePackageAction ?? false;
 }
 
+export function hasEndpointExceptionsPrivilege(
```
As discussed offline: I think the logic to determine if Lists for endpint exceptions is allowed should be moved to the current module that detemines list access for the Detection engine (or somewhere similar) - here: 

https://github.com/elastic/kibana/blob/6efef0496077f4e61d49e4e43f75941ff3d98d9e/x-pack/plugins/security_solution/public/detections/components/user_privileges/use_fetch_list_privileges.ts#L25

I'm ok if you want to leave it as is for now and refactor later.

paul-tavares replied:
Any feedback on this one or are still working on it?

ashokaditya replied:
@paul-tavares Still working on this. There are still some missing pieces that need to be in place to restrict APIs on Serverless and not on ESS. Once everything is working with this approach I'll reach out to you and discuss the re-factor. I do agree that this should move out of here.

paul-tavares commented on x-pack/plugins/security_solution/server/lists_integration/endpoint/validators/base_validator.ts:
```diff
@@ -74,6 +77,22 @@ export class BaseValidator {
+      AppFeatureSecurityKey.endpointExceptions
+    );
+
+    // for serverless only when Endpoint Exceptions is not enabled
+    // verify that the user has the privilege to create/read/update/delete endpoint exceptions
+    if (!isEndpointExceptionsFeaturesEnabled && !(await this.endpointAuthzPromise)[privilege]) {
+      throw new EndpointExceptionsValidationError('Endpoint exceptions authorization failure', 403);
+    }
```
Not really a fan of this should being done here. It can lead to these types of logic changes happening in multiple places.

It would be best if the Authz for exceptions (`canReadEndpointExceptions`, `canWriteEndpointExceptions`) were defined in a way that they already took into consideration the appFeature (just like we do today when we check Kbn license levels).  Maybe `appFeatues` should be pass into out `authz.ts` during calculation?

semd replied:
Yes, this conditional is not correct, we should only check the capabilities here: 

```
if (!(await this.endpointAuthzPromise)[privilege]) {
```

As per the configuration of the `endpointExceptions` AppFeatureKey (in [ess](https://github.com/elastic/kibana/blob/1f7a383b7becc94d8a13b8b2163635f3bf204d02/x-pack/plugins/security_solution_ess/server/app_features/security_app_features_config.ts#L48-L50) and in [serverless](https://github.com/elastic/kibana/blob/1f7a383b7becc94d8a13b8b2163635f3bf204d02/x-pack/plugins/security_solution_serverless/server/app_features/security_app_features_config.ts#L41-L43)), we can check those capabilities (via endpoint authz module in your case), and that's all we need to do: 

- If the `AppFeatureSecurityKey.endpointExceptions` is enabled: the _showEndpointExceptions_ and _crudEndpointExceptions_ capabilities will be added to the Security Solution privileges, and they will be granted depending on the user role.

- If the `AppFeatureSecurityKey.endpointExceptions` is not enabled: the _showEndpointExceptions_ and _crudEndpointExceptions_ capabilities will never be added to the Security Solution privileges, so they will never be granted, regardless of the user role.

@paul-tavares In the `authz.ts`, the appFeatures keys are already taken into consideration when checking the capabilities, it works in a declarative way using AppFeaturesConfigs linked above, there's no need to do explicit checks to the appFeatures in the authz code.

When validating privileges, as a general rule of thumb to not overcomplicate things, we can ask ourselves:

Do we have access to the authz module (or Kibana capabilities directly)?
    - Yes: Check them, that's all we need to do.
    - No: Get the AppFeaturesService and check if the key is enabled (knowing that this method will not take role privileges into account)

ashokaditya replied:
02eb0a28ea613c5d308852eb926176f2b401b91a

paul-tavares reviewed (approved):
Reviewed only the Workload Defend team files and left some comments. I'll wait to hear back from you before 👍 it

ashokaditya committed:
Show upsell on endpoint exceptions tab

Display an upsell message when endpoint exceptions tab is forced loaded on Rule > Endpoint Security > details via a link

refs elastic/security-team/issues/7223

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

semd commented on x-pack/packages/security-solution/upselling/service/index.ts:
```diff
@@ -10,4 +10,5 @@ export type {
   SectionUpsellings,
   UpsellingSectionId,
   UpsellingMessageId,
+  MessageUpsellings,
```
Is this export needed? It does not seem to be used

ashokaditya replied:
It is not used in this PR but it is a missing import from a [previous PR](https://github.com/elastic/kibana/pull/163406). I was seeing a type error in my PR cause of it.

semd commented on x-pack/plugins/security_solution_serverless/public/upselling/register_upsellings.tsx:
```diff
@@ -124,9 +133,14 @@ export const upsellingSections: UpsellingSections = [
     component: EndpointPolicyProtectionsLazy,
   },
+  {
+    id: 'endpointExceptions',
```
I was confused by this one, the ID looks too generic. If it is only needs to be displayed in the "exceptions" section of the Rules details page, it would be better to use a more specific ID:
```suggestion
    id: 'ruleDetailsEndpointExceptions',
```

ashokaditya replied:
52d0bc9b88c0016b64f2adf6e11846ebd8b0ab66

semd commented on x-pack/plugins/security_solution/public/rules/links.ts:
```diff
@@ -78,6 +78,7 @@ export const links: LinkItem = {
       }),
       landingIcon: IconConsoleCloud,
       path: EXCEPTIONS_PATH,
+      capabilities: [`${SERVER_APP_ID}.showEndpointExceptions`],
```
💯

semd commented on x-pack/plugins/security_solution/server/lists_integration/endpoint/validators/base_validator.ts:
```diff
@@ -74,6 +77,22 @@ export class BaseValidator {
+  protected async validateHasEndpointExceptionsPrivileges(
+    privilege: keyof EndpointAuthz
+  ): Promise<void> {
+    const appFeaturesService = this.endpointAppContext.getAppFeaturesService();
+
+    const isEndpointExceptionsFeaturesEnabled = (await appFeaturesService).isEnabled(
+      AppFeatureSecurityKey.endpointExceptions
+    );
```
Why is the `this.endpointAppContext.getAppFeaturesService()` an `async` function? The `AppFeaturesService` does not need to do any async logic, it is available synchronously.

juliaElastic reviewed (approved):
Fleet changes LGTM

ashokaditya committed:
rename

review suggestions @semd

ashokaditya committed:
only check capabilities and not app feature

review changes @paul-tavares @semd

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

semd commented on x-pack/plugins/security_solution/public/detection_engine/rule_details_ui/pages/rule_details/use_rule_details_tabs.tsx:
```diff
@@ -80,9 +82,17 @@ export const useRuleDetailsTabs = ({
 
+  const { loading: listsConfigLoading, needsConfiguration: needsListsConfiguration } =
+    useListsConfig();
+  const canReadEndpointExceptions = useHasSecurityCapability('crudEndpointExceptions');
```
This looks a bit inconsistent, but I am not sure if this should be:
`const canReadEndpointExceptions = useHasSecurityCapability('showEndpointExceptions');`
or
`const canWriteEndpointExceptions = useHasSecurityCapability('crudEndpointExceptions');`

ashokaditya replied:
Good catch. I think I missed some renaming here. It is indeed checking `crudEndpointExceptions` here.

ashokaditya replied:
dc4bb74cab9ad1223f67b4111961c44e3efe6afa

ashokaditya committed:
rename

review changes @semd

ashokaditya committed:
remove redundant method for now

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
fix merge duplicate imports

ashokaditya committed:
fix privilege calculations for endpoint expections

refs 02eb0a28ea613c5d308852eb926176f2b401b91a

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

ashokaditya committed:
fix types

refs 72fdb540e7c11ed27922b14552ae06707ce12e11

ashokaditya committed:
unit tests for `getAuthorizationFromPrivileges`

refs 72fdb540e7c11ed27922b14552ae06707ce12e11

ashokaditya committed:
fix types

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

paul-tavares commented on x-pack/plugins/security_solution_serverless/public/upselling/sections/rule_details_endpoint_exceptions/index.tsx:
```diff
@@ -0,0 +1,14 @@
+/*
```
same here. If our team is owner, please move it to the existing `endpoint_management` folder

ashokaditya replied:
a9751c61a3cbde0a7c9a8da5e764b8a482003b36

paul-tavares commented on x-pack/plugins/security_solution_serverless/public/upselling/sections/rule_details_endpoint_exceptions/rule_details_endpoint_exceptions.tsx:
```diff
@@ -0,0 +1,62 @@
+/*
```
same :)

paul-tavares commented on x-pack/plugins/fleet/common/authz.test.ts:
```diff
@@ -10,8 +10,11 @@ import { DEFAULT_APP_CATEGORIES } from '@kbn/core-application-common';
 import { TRANSFORM_PLUGIN_ID } from './constants/plugin';
 
 import {
+  calculateEndpointExceptionsPrivilegesFromCapabilities,
```
I forgot to ask this before: but were fleet changes really necessary for the endpoint exceptions authz?

We normally only have to touch fleet authz for when one of our kibana privileges needs to be used in fleet to allow access to one of the APIs. (right?)

ashokaditya replied:
I believe it was necessary to add this to fleet authz same as we did for endpoint package privileges, as privilege verification is done via fleetAuthz for all endpoint/exception_list artifact APIs.

paul-tavares replied:
I don't know about that. We only run our privileges through fleet so that we can enable fleet APIs for kibana features controls that are applicable to endpoint management. 

For endpoint exceptions, I don't think they need access to fleet APIs (ex. they don't support the "per policy" functionality), so it should all have been done from within Security Solution? (ex. that's how it works today, prior to RBAC being introduced for serverless).

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
add endpoint exceptions to endpoint complete

refs elastic/security-team/issues/7223

ashokaditya committed:
tidy up a bit

ashokaditya committed:
fix typo

ashokaditya committed:
update codeowners file

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

semd commented on x-pack/plugins/security_solution_serverless/public/upselling/register_upsellings.tsx:
```diff
@@ -40,7 +44,7 @@ interface UpsellingsMessageConfig {
 }
 
-type UpsellingPages = Array<UpsellingsConfig & { pageName: SecurityPageName }>;
+type UpsellingPages = Array<UpsellingsConfig & { pageName: SecurityPageName; path?: string }>;
```
Is this needed?

ashokaditya replied:
Good catch! Not needed at all. I think I added it for some reason and then forgot to clean it up

ashokaditya replied:
b7cbf121e1fe70d8a8f35d83e951386a04f18866

semd reviewed (approved):
LGTM! 
Great work!! 🚀

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
cleanup

review changes @semd

ashokaditya committed:
Merge branch 'task/dw-gate-endpoint-exceptions-via-serverless-pli-7223' of github.com:ashokaditya/kibana into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
fix merge error

ashokaditya commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

michaelolo24 reviewed (approved):
LGTM, thanks!

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

kibanamachine committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

kevinlog commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
fix run time error

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
tidy up

refs 141a413f72a85a4d34361b108a9a3f423e0b3dd7

ashokaditya committed:
show exceptions tab for sub feature role with read only permission

for serverless users with customised sub feature roles with `endpoint_exceptions_read` role show endpoint exceptions tab

ashokaditya committed:
disable actions for sub feature role

For serverless user with sub feature read only access to endpoint exceptions, disable edit/delete actions

refs elastic/security-team/issues/7223

ashokaditya committed:
disable edit/delete on rule details tabs

similar to 7cffc0e81772a5b8059e0737218eb533e2d8dc7a

refs elastic/security-team/issues/7223

banderror commented on x-pack/plugins/security_solution/public/detection_engine/rule_details_ui/pages/rule_details/hooks/use_get_endpoint_exceptions_unavailablle_component.tsx:
```diff
@@ -0,0 +1,13 @@
+ * 2.0.
+ */
+
+import type React from 'react';
+import { useUpsellingComponent } from '../../../../../common/hooks/use_upselling';
+
+export const useGetEndpointExceptionsUnavailableComponent = (): React.ComponentType | null => {
+  return useUpsellingComponent('ruleDetailsEndpointExceptions');
+};
```
I'd like to see if we can improve code ownership for the changes made to the Rule Details page in this PR.

Can we have somewhere an `endpoint_exceptions` folder that would be owned by you folks? Options off the top of my head:

- `x-pack/plugins/security_solution/public/detection_engine/rule_details_ui/components/endpoint_exceptions`
- `x-pack/plugins/security_solution/public/detection_engine/rule_details_ui/pages/rule_details/endpoint_exceptions`
- `x-pack/plugins/security_solution/public/detection_engine/endpoint_exceptions/components`

This `endpoint_exceptions` folder would be owned by @elastic/security-defend-workflows and would be exporting components and hooks for use on the Rule Details and optionally other pages of the Detection Engine.

The `useGetEndpointExceptionsUnavailableComponent` hook would live under the `endpoint_exceptions` folder.

ashokaditya replied:
refactored 35d0a11794d946f21633c0770901b13054bc90ff

banderror reviewed (commented):
Hey @ashokaditya, thank you for implementing these PLI checks. From the Rule Management side, we don't really own the exceptions functionality, so I was primarily looking into the diff and overall code of the Rule Details page. I have a few suggestions for relatively simple refactoring that could improve code ownership for endpoint exceptions.

I will do some basic testing based on the PR description, thank you for providing these instructions.

ashokaditya committed:
unify capability check logic in a hook

ashokaditya committed:
move file

review changes @banderror

ashokaditya commented on x-pack/plugins/security_solution/public/detection_engine/rule_details_ui/pages/rule_details/use_rule_details_tabs.tsx:
```diff
@@ -80,9 +82,17 @@ export const useRuleDetailsTabs = ({
+  const canReadEndpointExceptions = useMemo(
+    () => !listsConfigLoading && !needsListsConfiguration && canSeeEndpointExceptions,
+    [canSeeEndpointExceptions, listsConfigLoading, needsListsConfiguration]
+  );
```
Some refactor that I had in mind 6e3a96c24d60c5c156bede882d21ad229bd39ce2. I'll get to the rest of it in a new PR

ashokaditya replied:
refactored it here 35d0a11794d946f21633c0770901b13054bc90ff

ashokaditya commented on x-pack/plugins/security_solution/public/detection_engine/rule_details_ui/pages/rule_details/hooks/use_get_endpoint_exceptions_unavailablle_component.tsx:
94d78dde913f1754247e47d46aa2a73560465af9

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

banderror commented:
I did some testing of the changes on the Rule Details page and it LGTM 👍 

@ashokaditya I wasn't able to run Serverless Kibana using the provided instructions in the description. That's probably my lack of knowledge of how to run Kibana in Serverless. I had to comment out everything in `kibana.dev.yml` and log in as `elastic_serverless`. How do we login as `serverless_security`?

The Endpoint exceptions tab on the Rule Details page worked fine with the following config:

```yml
xpack.securitySolutionServerless.productTypes:
  [
    { product_line: 'security', product_tier: 'complete' },
    # { product_line: 'endpoint', product_tier: 'complete' },
  ]
```

I was able to open this "tab" using its direct URL, but it showed that I should upgrade to Endpoint Essentials. The tab itself was hidden 👍 

<img width="1792" alt="Screenshot 2023-09-29 at 19 53 28" src="https://github.com/elastic/kibana/assets/7359339/ce4c54b7-1c12-4fd8-8f5a-6d756c334680">

Nit: the UI is slightly different for the endpoint list page

<img width="1792" alt="Screenshot 2023-09-29 at 19 55 06" src="https://github.com/elastic/kibana/assets/7359339/b72548ad-020e-4f56-abcf-51578a72f1a8">

ashokaditya committed:
refactor

review suggestions @banderror

ashokaditya commented on x-pack/plugins/security_solution/public/detection_engine/rule_details_ui/pages/rule_details/index.tsx:
```diff
@@ -780,13 +783,21 @@ const RuleDetailsPageComponent: React.FC<DetectionEngineComponentProps> = ({
+                        <EndpointExceptionsUnavailableComponent />
+                      </EuiFlexItem>
+                    </EuiFlexGroup>
+                  )}
```
refactored 35d0a11794d946f21633c0770901b13054bc90ff

ashokaditya commented:
Thanks for the review @banderror. I've managed to refactor as you suggested. Here are some thoughts on testing locally.

> I did some testing of the changes on the Rule Details page and it LGTM 👍
> 
> @ashokaditya I wasn't able to run Serverless Kibana using the provided instructions in the description. That's probably my lack of knowledge of how to run Kibana in Serverless. I had to comment out everything in `kibana.dev.yml` and log in as `elastic_serverless`. How do we login as `serverless_security`?

You tested it the correct way @banderror. `elastic_serverless` is actually `superuser` so one needs to test APIs with custom user/roles (which is still being worked on by the core team to allow devs to do this easily).  

For now, I'm testing this locally by 
1. Running ES snapshot/ESS kibana, 
2. Add Endpoint Security rule
3. Add some endpoint exceptions using exception_list API with `elastic:changeme`,
4. Creating roles/users using APIs (`PUT api/security/role/:rolename`, `POST /internal/security/users/:username`) and then, while the ES snapshot is running,
5. I switch to serverless kibana with `yarn serverless-security --no-base-path` and the log in with the created custom user(with the custom role). 

Note: To be able to do the above switch one also needs to comment out [this block](https://github.com/elastic/kibana/blob/d4b7a480e69fa57c0f359b6ea3493627c4f1316d/src/cli/serve/serve.js#L110)

I know this is tedious for manual testing, but it works reliably for now. Feel free to try it out if you like.

> The Endpoint exceptions tab on the Rule Details page worked fine with the following config:
> 
> ```yaml
> xpack.securitySolutionServerless.productTypes:
>   [
>     { product_line: 'security', product_tier: 'complete' },
>     # { product_line: 'endpoint', product_tier: 'complete' },
>   ]
> ```
Superb 👍🏼.  

> I was able to open this "tab" using its direct URL, but it showed that I should upgrade to Endpoint Essentials. The tab itself was hidden 👍

Excellent. This is expected.

> Nit: the UI is slightly different for the endpoint list page

Quite right! We will update this later along with the text info. These changes only affect the serverless UX anyway.

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

banderror reviewed (approved):
Thank you for addressing my comments @ashokaditya, LGTM 👍

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
fix type check error

refs elastic/kibana/pull/166781

ashokaditya committed:
use defined const

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
Merge branch 'main' into task/dw-gate-endpoint-exceptions-via-serverless-pli-7223

ashokaditya committed:
fix merge error
